### 1. Write a Python/Numpy code block that finds the distinct/unique common items between these two arrays:



```
a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])
```

Your output should contain only the distinct overlapping values. For example, if a 2 is found in both array ​__a__ ​and array ​__b__​, your output should contain only one 2 even if array ​__a__ ​contains more than one 2 within it.



In [35]:
import numpy as np

a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])

print(np.intersect1d(a, b))

[2 4]


### 2. Create the following 5x3 array using knowledge you have of Python’s / NumPy’s sequencing functionality so that you do not need to explicitly key in every integer value.

|      |      |      |
| :--: | :--: | :--: |
| 1    | 6    | 11   |
| 2    | 7    | 12   | 
| 3    | 8    | 13   |
| 4    | 9    | 14   |
| 5    | 10   | 15   | 

In [36]:
array = np.arange(1, 16).reshape(3, 5)
array = np.transpose(array)
array

array([[ 1,  6, 11],
       [ 2,  7, 12],
       [ 3,  8, 13],
       [ 4,  9, 14],
       [ 5, 10, 15]])

### 3. The process of transforming a multidimensional array into a unidimensional array is referred to as“flattening”. Transform the 5x3 array shown above in Problem 2 into a unidimensional array such that the sequence of values contained within the array is as follows: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15

In [37]:
flattened_array = np.transpose(array).flatten()
flattened_array

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

### 4. Transform the 2-D array shown in Problem 2 into a 3 dimensional array such that the first column becomes the first dimension of the 3-D array, the second column becomes the second dimension of the 3-D array, and the third column becomes the third dimension of the 3-D array

In [38]:
array_3d = np.transpose(array).reshape(3, 5, -1)
array_3d

array([[[ 1],
        [ 2],
        [ 3],
        [ 4],
        [ 5]],

       [[ 6],
        [ 7],
        [ 8],
        [ 9],
        [10]],

       [[11],
        [12],
        [13],
        [14],
        [15]]])

### 5. Transform the 3-D array you created in Problem 4 back to the 2-dimensional format shown in Problem 2

In [39]:
array = np.transpose(array_3d.reshape(3, 5))
array

array([[ 1,  6, 11],
       [ 2,  7, 12],
       [ 3,  8, 13],
       [ 4,  9, 14],
       [ 5, 10, 15]])

### 6. You are given the following two arrays:



```
a = np.array([12, 5, 7, 15, 3, 1, 8])
b = np.array([14, 6, 3, 11, 19, 12, 5])

```

Write a Python/Numpy code block that removes from array ​__a__ ​any items that are also present in array __b__



In [40]:
a = np.array([12, 5, 7, 15, 3, 1, 8])
b = np.array([14, 6, 3, 11, 19, 12, 5])
a = np.array([n for n in a if n not in b])
print(a)

[ 7 15  1  8]


### 7. ​For this problem you will load a small data set to your online GitHub repository and then read it into a NumPy 2D array. You will then use the content that NumPy array to answer a series of analytical questions.

In [41]:
import pandas

# supress scientific notation
np.set_printoptions(suppress = True)

# read the data file from GitHub and convert to a NumPy ndarray object
nyc_water = pandas.read_csv('https://raw.githubusercontent.com/thesabermaniac/NumpyAssignment/main/cityData.csv').to_numpy()

# verify you have an ndarray object
type(nyc_water)

numpy.ndarray

#### What is the maximum yearly NYC consumption of water in millions of gallons per day?

In [61]:
max_gpd = np.max(nyc_water[:,2], axis=0)
max_gpd

1512.0

The max yearly consumption of water in nyc is 1,512 million gallons per day (or 1,512,000,000)

#### How many calendar years are represented within this data set? NumPy's ​shape ​command is one way to find out.

In [63]:
num_of_years = nyc_water[:,0].max() - nyc_water[:,0].min()
num_of_years

38.0

There are 38 years represented in the dataset

#### What is the mean and the standard deviation of the per capita daily water consumption?

In [65]:
mean_per_capita_water = nyc_water[:,3].mean()
print(mean_per_capita_water)

std_per_capita_water = nyc_water[:,3].std()
print(std_per_capita_water)

160.30769230769232
31.51894330370624


Mean = 160.308

Standard Deviation = 31.519

#### What is the increase or decrease in population from year to year? Use NumPy's `​diff` ​function to create an array  of differences and save that to a variable called "​pop_diff​", then print that variable to the screen.

In [60]:
np.diff(nyc_water[:, 1], axis=0)

array([-30461. ,  17602. ,  19864. ,  72119. ,  53290. ,  39540. ,
        45192. ,  23230. ,  11243. ,  -9544. ,  -8525. ,  38851. ,
        54443. ,  77222. ,  64292. ,  62582. ,  64772. ,  75631. ,
        84816. ,  89401. ,  60618. ,  16685.5,  16685.5,  16685.5,
        16685.5,  16685.5,  16685.5,  16685.5,  16685.5,  16685.5,
        16685.5, 108965. ,  77081. ,  61281. ,  49530. ,  44512. ,
        21171. ,  85025. ])